File: src/commands/moderation/timeout.js
const { SlashCommandBuilder, PermissionsBitField } = require('discord.js');

module.exports = {
    data: new SlashCommandBuilder()
        .setName('timeout')
        .setDescription('Times out a user')
        .setDefaultMemberPermissions(PermissionsBitField.Flags.ModerateMembers)
        .addUserOption(option => option.setName('target').setDescription('The user to time out').setRequired(true))
        .addStringOption(option => option.setName('time').setDescription('The duration to time the user out').setRequired(true).addChoices(
            { name: '60 seconds', value: '60' },
            { name: '5 minutes', value: '300' },
            { name: '10 minutes', value: '600' },
            { name: '30 minutes', value: '1800' },
            { name: '1 hour', value: '3600' },
            { name: '12 hours', value: '43200' },
            { name: '1 day', value: '86400' },
            { name: '1 week', value: '604800' },
            { name: '1 month', value: '2629743' }
        ))
        .addStringOption(option => option.setName('reason').setDescription('The reason for the timeout').setRequired(false)),
    async execute(interaction) {
        const user = interaction.options.getUser('target');
        const member = await interaction.guild.members.fetch(user.id).catch(console.error);
        let reason = interaction.options.getString('reason');
        let time = interaction.options.getString('time');
        if (!time) time = '60';
        if (!reason) reason = 'No reason provided';

        if (!interaction.member.permissions.has(PermissionsBitField.Flags.ModerateMembers)) return await interaction.reply({ content: 'You do not have permission to timeout this user', ephemeral: true })
        if (!member.kickable) return await interaction.reply({ content: 'This user cannot be timed out', ephemeral: true })
        if (!member) return await interaction.reply({ content: `User ${user.tag} is not in this server`, ephemeral: true })
        if (interaction.member.id === user.id) return await interaction.reply({ content: 'You cannot timeout yourself', ephemeral: true })
        if (member.permissions.has(PermissionsBitField.Flags.Administrator)) return await interaction.reply({ content: 'You cannot timeout this user', ephemeral: true })

        await member.timeout(time * 1000, reason).catch(console.error);
        await interaction.reply({
            content: `Timed out ${user.tag} for ${time} seconds for ${reason}`,
            ephemeral: true
        })
    }
};

File: src/commands/moderation/timeout.rs
use crate::types::Context;
use color_eyre::Result;
use poise::serenity_prelude::all::{EditMember, Timestamp, User};
use regex::Regex;

#[poise::command(slash_command, guild_only, required_permissions = "MODERATE_MEMBERS")]
pub async fn timeout(
    ctx: Context<'_>,
    user: User,
    #[description = "how long the time out is for. (5s, 2m, 12h, 3d, 2w)"] duration: String,
    reason: Option<String>,
) -> Result<()> {
    let guild = ctx
        .serenity_context()
        .http
        .get_guild(ctx.guild_id().unwrap())
        .await?;

    let timeout_max_value = 2419200;

    let user_member = guild.member(ctx, user.id).await.unwrap();
    let bot_member = guild.member(ctx, ctx.framework().bot_id).await.unwrap();
    let author_member = guild.member(ctx, ctx.author()).await.unwrap();

    let user_highest_role = user_member
        .roles
        .iter()
        .filter_map(|role| guild.roles.get(role))
        .max_by_key(|role| role.position);

    let author_highest_role = author_member
        .roles
        .iter()
        .filter_map(|role| guild.roles.get(role))
        .max_by_key(|role| role.position);

    let bot_highest_role = bot_member
        .roles
        .iter()
        .filter_map(|role| guild.roles.get(role))
        .max_by_key(|role| role.position);

    if guild
        .user_permissions_in(
            &ctx.guild_channel().await.unwrap(),
            &guild.member(ctx, user.id).await.unwrap(),
        )
        .administrator()
    {
        ctx.say("Target user has permission: ``Administrator``")
            .await?;
        return Ok(());
    }

    if !guild
        .user_permissions_in(
            &ctx.guild_channel().await.unwrap(),
            &guild.member(ctx, ctx.framework().bot_id).await.unwrap(),
        )
        .moderate_members()
    {
        ctx.say("Bot missing permission: ``Moderate Members``")
            .await?;
        return Ok(());
    }

    if author_highest_role < user_highest_role {
        ctx.say("User has higher role then you.").await?;
        return Ok(());
    }

    if bot_highest_role < user_highest_role {
        ctx.say("User has higher role then bot!").await?;
        return Ok(());
    }

    let Some(seconds) = parse_duration(&duration) else {
        ctx.say("Invalid duration! (Valid durations: 5s, 2m, 12h, 3d, 2w)")
            .await?;
        return Ok(());
    };

    if seconds > timeout_max_value {
        ctx.say("Duration too long! (Max duration: 28d)").await?;
        return Ok(());
    }

    let seconds = i64::try_from(seconds);
    if seconds.is_err() {
        ctx.say("Duration value is too large!").await?;
        return Ok(());
    }

    // safer unwrap then cloudflare lol
    let timeout_until =
        Timestamp::from_unix_timestamp(Timestamp::now().unix_timestamp() + seconds.unwrap())?;

    guild
        .member(ctx, user.id)
        .await?
        .edit(
            ctx,
            EditMember::new()
                .disable_communication_until(timeout_until.to_string())
                .audit_log_reason(&reason.unwrap_or("No reason provided.".to_string())),
        )
        .await?;

    ctx.say(format!("Timed out <@{}> for {}!", user.id, duration))
        .await?;

    Ok(())
}

fn parse_duration(input: &str) -> Option<u64> {
    let re = Regex::new(r"(\d+)([smhdw])").unwrap();

    if let Some(caps) = re.captures(input) {
        let value: u64 = caps[1].parse().ok()?;
        let unit = &caps[2];

        let seconds = match unit {
            "s" => value,
            "m" => value * 60,
            "h" => value * 3600,
            "d" => value * 86400,
            "w" => value * 604800,
            _ => return None,
        };

        Some(seconds)
    } else {
        None
    }
}
